File: ruby/029.rb
# 029
# Divide Two Integers
# 2015-05-04
#####################################################
# @param {Integer} dividend
# @param {Integer} divisor
# @return {Integer}
def divide(dividend, divisor)
  return 0 if divisor == 0
  sign = ((dividend > 0) ^ (divisor > 0))
  dividend = -dividend if dividend < 0
  divisor = -divisor if divisor < 0

  int_max = 2147483647
  d = [divisor]
  i = 1
  while d[i-1] < dividend and d[i-1] < int_max
    d[i] = d[i-1] + d[i-1]
    i+=1
  end
  i-=1
  res = 0
  while dividend >= divisor
    while d[i] > dividend and i > 0
      i -= 1
    end
    r = 1
    for j in (0...i)
      r += r
    end
    res += r
    dividend -= d[i]
  end
  if sign
    res = -res
  end
  [res, int_max].min
end

#####################################################

INT_MAX = 2**31-1
INT_MIN = -2**31

dividend = INT_MIN
divisor = -1

puts "#{dividend} / #{divisor} = #{divide(dividend, divisor)}"

File: c/029.c
//
//  029.c
//  leetcode
//
//  Created by Hang Zhang on 5/4/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <stdio.h>
#include <limits.h>
#include <stdbool.h>

int divide(int dividend, int divisor) {
    if (divisor == 0)
        return 0;
    long dividend2 = dividend;
    long divisor2 = divisor;
    bool sign = dividend2 > 0 ^ divisor2 > 0;
    if (dividend2 < 0)
        dividend2 = -dividend2;
    if (divisor2 < 0)
        divisor2 = -divisor2;

    int i;
    long d[32];
    d[0] = divisor2;
    for (i = 1; d[i-1] < dividend2 && d[i-1] < INT_MAX; i++)
        d[i] = d[i-1] + d[i-1];
    i--;
    long res = 0;
    while (dividend2 >= divisor2) {
        while (d[i] > dividend2 && i > 0)
            i--;
        long r = 1;
        for (int j = 0; j < i; j++)
            r += r;
        res += r;
        dividend2 -= d[i];
    }
    if (sign)
        res = -res;
    return res > INT_MAX ? INT_MAX : (int)res;
}

int main(int arg, char *argv[]) {
    // insert code here...
    printf("LeetCode 029 Divide Two Integers, C ...\n\n");

    int dividend = INT_MIN;
    int divisor = -1;
    printf("%d / %d = %d\n", dividend, divisor, divide(dividend, divisor));

    return 0;
}